std/random.d

```diff
@@ -564,62 +572,135 @@ Parameters for the generator.
     /// The default seed value.
     enum UIntType defaultSeed = 5489u;
 
-/**
-   Constructs a MersenneTwisterEngine object.
-*/
+    /// Collection of all state variables
+    /// used by the generator
+    private struct State
+    {
+        private UIntType y = void;
+        private UIntType z = 0;
+        private UIntType index = void;
+        private UIntType[n] data = void;
+
+    }
+
+    /// State variables used by the generator;
+    /// initialized to values equivalent to
+    /// explicitly seeding the generator with
+    /// `defaultSeed`
+    private State state = defaultState();
+    // NOTE: the above is a workaround to ensure
+    // backwards compatibility with the original
+    // implementation, which permitted implicit
+    // construction.  With `@disable this();`
+    // it would not be necessary.
+
+    /**
+       Constructs a MersenneTwisterEngine object.
+    */
     this(UIntType value) @safe pure nothrow @nogc
     {
-        seed(value);
+        this.seed(value);
     }
 
-/**
-   Seeds a MersenneTwisterEngine object.
-   Note:
-   This seed function gives 2^32 starting points. To allow the RNG to be started in any one of its
-   internal states use the seed overload taking an InputRange.
-*/
-    void seed()(UIntType value = defaultSeed) @safe pure nothrow @nogc
+    /**
+       Constructs a MersenneTwisterEngine object using a range
+       to seed the generator, which must have at least as many
+       elements as `data`.
+    */
+    this(T)(T range)
+        if (isInputRange!T && is(Unqual!(ElementType!T) == UIntType))
+    {
+        this.seed(range);
+    }
+
+    /**
+       Constructs a MersenneTwisterEngine object by
+       duplicating the state of an existing instance
+    */
+    this(ref typeof(this) that)
```

Comment:
A small note on the motivation for this constructor overload (which I should possibly add to the patch description if we decide to keep it).  A long-term goal is to remove the `.save` property of Phobos' random number generators: however, it is still useful to be able to deliberately deep-copy the state of an RNG in order to create a second generator that will produce identical results.

Having a constructor that copies an existing generator provides a way to "wrap up" the details of duplicating a generator in a way that can both be used directly (to construct a new instance) and also wrapped by other methods that might require it (`save` while it's still needed to avoid breaking change; `dup` has been proposed in future as a method that provides duplication without creating a forward range).

However, right now, this is creating a new feature that we might not want to support.  So it might be better to either make this constructor instance `private` or to remove it for now and just rework the internals of `save`.